Implement `count` with SSE 4.2 and AVX2

0xd34df00d opened the pull request:
This provides a fairly significant performance boost: `count`ing lines via `BS.count 10` was taking about 1.1 s with an `mmap`ed 1.8 gigabyte file, and now it takes 0.35 s. For comparison, `length . BS.lines` takes about 0.4 s (IMO `count` should definitely be at least as fast), and Unix `wc` takes about 0.3-0.4 s.

sjakobi commented:
Interesting!

I'm wondering why `count` didn't use `memchr` already while other functions do:

https://github.com/haskell/bytestring/blob/95fe6bdf13c9cc86c1c880164f7844d61d989574/Data/ByteString.hs#L1001-L1013

https://github.com/haskell/bytestring/blob/95fe6bdf13c9cc86c1c880164f7844d61d989574/Data/ByteString.hs#L1085-L1089

https://github.com/haskell/bytestring/commit/c29225bb3c4b76241f735795c44d3f513098f894 mentions using `memchr` for `count` but goes with a different implementation for some reason?!

It would be nice if you could share your benchmark but I guess that's somewhat blocked on #196.

https://github.com/haskell/bytestring/issues/109 also looks related.

0xd34df00d commented:
My benchmark is something really silly, along the lines of
```haskell
import qualified Data.ByteString as BS
import System.IO.Posix.MMap

main :: IO ()
main = do
  contents <- unsafeMMapFile "/huge/file/on/tmpfs.txt"
  print $ BS.count 10 contents
```

The file size is about 2 gigs, and it's on a `tmpfs` partition with no swapping, so no IO involved. Since the file is `mmap`ed, there's also no GC pressure at all.

Thanks for the pointer to #109 , I missed it! While I'm at this task I can just as well try to brush up my SIMD intrinsics skills and also implement a version using `PCMPESTRM` as suggested in one of the comments — it might indeed be beneficial, especially for strings where the character in question occurs often enough.

In case I get decent results, what's `bytestring`'s policy on SIMD, supported architectures and adding flags for those, or C compiler dependency? IIRC given code along the lines of (I don't remember the right attribute syntax off the top of my head though)
```c
__attribute__((target("sse42")))
void foo() { ... }

__attribute__((target("avx")))
void foo() { ... }

/* fallback */
void foo() { ... }
```
gcc can generate some dispatching code to select the right implementation at runtime according to the current (runtime) CPU architecture.

0xd34df00d commented:
Alright, did some experimenting with `PCMPESTRM`, and the results are quite in favour of it, especially given that its performance doesn't depend on the character frequency.

1. When looking for `'\n'` (with a frequency of about 0.8%), `memchr`-based run time is about 311 ms. With `PCMPESTRM` it's a bit (but statistically significantly) lower — about 295 ms.
2. When looking for `' '` that occurs exactly twice as often, `memchr` runtime degrades to 425 ms, while `PCMPESTRM` stays at 295 ms.
3. When looking for `,` with frequency of about 10% (my test file is a CSV), `memchr` is dying at about 1150 ms, while `PCMPESTRM` — you guessed it — stays at 295 ms.

So I guess if every other character would be the character in question, `memchr`-based solution would be a no-go.

My gut feel is that with `PCMPESTRM` the performance is limited by CPI, so I'll play around some more to see if using 256-bit-wide registers and instructions would help, so I'm not pushing what I currently have, but just for the reference, it's
```c
__attribute__((target("sse4.2")))
unsigned long fps_count(unsigned char *str, unsigned long len, unsigned char w) {
    __m128i pat = _mm_set1_epi8(w);

    const int mode = _SIDD_SBYTE_OPS | _SIDD_CMP_EQUAL_EACH;

    unsigned long res = 0;

    size_t i = 0;
    for (; i < len; i += 64) {
        __m128i p1 = _mm_lddqu_si128((const __m128i*)(str + i + 16 * 0));
        __m128i p2 = _mm_lddqu_si128((const __m128i*)(str + i + 16 * 1));
        __m128i p3 = _mm_lddqu_si128((const __m128i*)(str + i + 16 * 2));
        __m128i p4 = _mm_lddqu_si128((const __m128i*)(str + i + 16 * 3));
        __m128i r1 = _mm_cmpestrm(p1, 16, pat, 16, mode);
        __m128i r2 = _mm_cmpestrm(p2, 16, pat, 16, mode);
        __m128i r3 = _mm_cmpestrm(p3, 16, pat, 16, mode);
        __m128i r4 = _mm_cmpestrm(p4, 16, pat, 16, mode);
        res += _popcnt64(_mm_extract_epi64(r1, 0));
        res += _popcnt64(_mm_extract_epi64(r2, 0));
        res += _popcnt64(_mm_extract_epi64(r3, 0));
        res += _popcnt64(_mm_extract_epi64(r4, 0));
    }

    for (; i < len; ++i) {
        res += str[i] == w;
    }

    return res;
}
```

Let's probably start discussing how SIMD extensions-dependent things should be integrated?

sjakobi commented:
Wow, those results with `PCMPESTRM` look very promising!

> Let's probably start discussing how SIMD extensions-dependent things should be integrated?

Indeed – I can't be of much help there though! :)

Ping @hvr, @dcoutts!

@0xd34df00d You can also try garnering more feedback on the libraries mailing list.

0xd34df00d commented:
@sjakobi spent some more time at this, and AVX2 looks even better (see the table at https://github.com/0xd34df00d/counting-chars)!

I made some effort to make sure this still works on machines that don't have AVX2 (or even SSE4.2), but I'm not sure about the treatment of compilers other than gcc and clang, so ideally some other pair of eyes better suited for this sort of stuff than me would take a look at this.

Bodigrim commented:
@0xd34df00d this is a very inspiring result! Could you please raise a question about SIMD policies on [Libraries maillist](https://mail.haskell.org/cgi-bin/mailman/listinfo/libraries)?

cartazio commented:
THIS IS REAALLY COOL DEADBEEF, though of course i need to read it much close :)

cartazio commented:
for what its worth, theres no simd optimization guidelines for cbits parts of haskell libraries, and the main bottleneck /engineering challenge is having the right dynamic code selection piece, which this PR seems to do!

one important question is always: whats the overhead of the dynamic dispatch on smaller inputs?  (eg, is it better to only do the simd when inputs are above some size threshold?) plus stuff about which cpu generations actually benefit (which is tricky!)

cartazio commented on cbits/fpstring.c:
```diff
@@ -72,19 +83,163 @@ unsigned char fps_minimum(unsigned char *p, unsigned long  len) {
+    return fps_count_naive(str, len, w);
+#else
+    if (len <= 1024) {
+        return fps_count_naive(str, len, w);
```
oh sweet, it does do the fallback for small inputs :)

cartazio commented:
your dynamic dispatch looks correct, in isolation, *but*, if i'm reading the gcc/clang docs about function multi versioning, i think you're either using it wrong or underusing it. Or i need to read/play with those more and understand the tricks myself.

reference urls:
https://gcc.gnu.org/wiki/FunctionMultiVersioning
https://llvm.org/devmtg/2014-10/Slides/Christopher-Function%20Multiversioning%20Talk.pdf
https://hannes.hauswedell.net/post/2017/12/09/fmv/
https://clang.llvm.org/docs/AttributeReference.html#cpu-dispatch

https://gcc.gnu.org/onlinedocs/gcc-10.1.0/gcc/Function-Multiversioning.html#Function-Multiversioning 

to be clear: my fuzzy understanding is that if the file was compiled in c++ mode, then the different attributes but same function name would do an auto select? (and here you're doing a dynamic dispatch via c either way, so they have deliberately different names).  --- also clang apparently has different syntax for the attribute stuff..

to be clear, i literally learned this was a feature today :) , so thats great!

cartazio commented:
derp https://releases.llvm.org/7.0.0/tools/clang/docs/AttributeReference.html#target-gnu-target is the clang analogue?

cartazio commented:
https://releases.llvm.org/10.0.0/tools/clang/docs/AttributeReference.html#target

0xd34df00d commented:
Yay, thank you @cartazio for taking a look at this!

Regarding the dynamic dispatch — yes, gcc surely can and will generate the dispatching code from these attributes. clang, on the other hand, doesn't seem to always do this — at least, in my experience I had to manually write a similar dispaching code about 3 years ago. Maybe things got better since then!

Of course, I'm all for dropping manual detection and dispatch (the less code, the better), but I'm not entirely comfortable doing so given the experience above.

Also, speaking of the overhead of dynamic dispatch — in any case it only happens once per OS/C thread (or what's the right Haskell-lingo for this), and then it's a matter of a single comparison of a thread-local function pointer against 0, which a CPU's branch predictor will hopefully very quickly learn. My primary reason for using the naive function for smaller inputs is to avoid all the overhead for unaligned prologue/epilogue handlers of the SIMD versions.

cartazio commented:
no, no, it looks fine, i'm just not often given the experience of seeing well written dispatching portable simd.

I think that you *dont* need to do the attribute annotation if you're doing the manual dispatch? otoh, i literally have zero experience with this stuff so not sure :) .

i think the naive fall back for smaller inputs is the right choice though, so that seems fine to me.

cartazio commented:
lets make sure the build goes green and this code works with both recent gcc and clang both! :)

sjakobi commented:
A good compatibility test would be to build this on GHC's CI infrastructure. I've meant to properly document this for a while, but essentially you

1. Fork GHC on https://gitlab.haskell.org/ghc/ghc
2. Update the `libraries/bytestring` submodule to point at your fork
3. Start a CI pipeline for your GHC branch at https://gitlab.haskell.org/username/ghc/pipelines

cartazio commented:
@0xd34df00d  some build errors you should fix ;) 
```

cbits/fpstring.c:214:9: error:
606     warning: implicit declaration of function ‘__get_cpuid_count’ [-Wimplicit-function-declaration]
607         if (__get_cpuid_count(7, 0, &eax, &ebx, &ecx, &edx)) {
608             ^
609    |
610214 |     if (__get_cpuid_count(7, 0, &eax, &ebx, &ecx, &edx)) {
611    |         ^
612cbits/fpstring.c: In function ‘select_fps_simd_impl’:
613
614cbits/fpstring.c:214:9: error:
615     warning: implicit declaration of function ‘__get_cpuid_count’ [-Wimplicit-function-declaration]
616         if (__get_cpuid_count(7, 0, &eax, &ebx, &ecx, &edx)) {
617             ^
618    |
619214 |     if (__get_cpuid_count(7, 0, &eax, &ebx, &ecx, &edx)) {

```

cartazio commented:
i think the build is erroring there, but idk :)

0xd34df00d commented:
@cartazio,

> I think that you dont need to do the attribute annotation if you're doing the manual dispatch?

That depends on the compiler flags like `-march` (I think gcc and clang behave exactly the same here). If gcc/clang generate code for, say, generic x86_64 machine, long story short, they don't know they are allowed to use SSE 4.2 or AVX, and they will error out.

Re the build error, that's interesting, I'll take a look! Might be a different standard lib with different intrinsics exposed. Are you building on mac by any chance? 

@sjakobi, thanks for the instructions, will do!

Bodigrim commented:
@0xd34df00d are you still interested in this PR?

Bodigrim commented:
* `__get_cpuid_count` requires `gcc-6` and must be guarded by `#if __GNUC__ > 5`.
* The patch does not compile on MacOS and FreeBSD with `argument to '__builtin_ia32_pcmpestrm128' must be a constant integer` failures. That's because `clang` is stricter than `gcc`. 

These are probably solvable issues, but such amount of conditional compilation makes me uncomfortable to proceed. We are not in a position to test `bytestring` build against all possible combinations of C compilers and platforms.

@vdukhovni what's your take?

0xd34df00d commented:
@Bodigrim sorry for long reply, this completely fell out of my attention span. I'm definitely still interested!

Frankly, neither am I comfortable with such amount of platforms that I cannot test this on (_especially_ for such an important and foundational library). Short-term, seems like the best way is to have a cabal flag, controlling whether the hand-written SIMD implementation is used for this function (and possibly others), and disabled by default.

What I'd like to see as an ideal long-term solution is this function implemented purely in terms of GHC's SIMD primops. In fact, <a href="https://github.com/0xd34df00d/counting-chars-ghc-simd/blob/master/src/Data/ByteString/CountByte.hs">I tried that</a>, but some things are missing in the set of available primops (namely, there's an `undefined` there in place of something, but I did that a while ago so I don't remember what exactly is missing off the top of my head).

Bodigrim commented:
> Short-term, seems like the best way is to have a cabal flag, controlling whether the hand-written SIMD implementation is used for this function (and possibly others), and disabled by default.

Besides usability concerns (switching a flag of `bytestring` will cause everything to rebuild, including half of boot libraries), I'm unsure about the intention. Users care less about performance on a local machine than in production, but that's exactly where they have much less control over environment. 

Supporting an additional cabal flag is very costly from maintenance perspective. More configurations to test, more potential build failures, more support. 

I'm not opposed to SIMD in `bytestring` in general. Could your patch be rewritten in a more platform-independent way, possibly sacrificing some performance gains? CC @ethercrow who worked on #310.

vdukhovni commented:
> * `__get_cpuid_count` requires `gcc-6` and must be guarded by `#if __GNUC__ > 5`.
> * The patch does not compile on MacOS and FreeBSD with `argument to '__builtin_ia32_pcmpestrm128' must be a constant integer` failures. That's because `clang` is stricter than `gcc`.
> 
> These are probably solvable issues, but such amount of conditional compilation makes me uncomfortable to proceed. We are not in a position to test `bytestring` build against all possible combinations of C compilers and platforms.
> 
> @vdukhovni what's your take?

I don't have very strong views here, I don't have any experience programming with such intrinsics.  What I can say is that given that this was intended to be merely a foot-in-the-door demo of the principle, without a strong motivating use case for the actual functionality, it should wait until someone is inspired enough to figure out a way of making it clean. Just passage of time might make the code more portable on future systems...

Bodigrim commented:
@0xd34df00d could you please rebase atop the latest `master`? Besides resolving a conflict, the latest `master` has a much better CI, which will facilitate further discussion.

ethercrow commented:
> I'm not opposed to SIMD in bytestring in general. Could your patch be rewritten in a more platform-independent way, possibly sacrificing some performance gains? CC @ethercrow who worked on #310.

I'd prefer cpuid based approach as opposed to cabal flags approach. This way the user doesn't have to think about or even be aware of CPU instruction sets.

If some compilers or versions of compilers don't support everything that's necessary then we can just extend the CPP condition around the naive version from `ifndef __x86_64__` to something like `if (not __x86_64__) || (GCC < 6) || (CLANG < 42)`, so that at least we don't have many code paths for many compilers.

Bodigrim commented:
> we can just extend the CPP condition around the naive version from `ifndef __x86_64__` to something like `if (not __x86_64__) || (GCC < 6) || (CLANG < 42)`

This way we need to test builds against an older gcc, a newer gcc, an older clang and a newer clang. This is four additional configurations out of nowhere. And they should be tested on two CI servers: one with AVX2 and another without - no idea where I can get them. Alternatively one can conjure up a flag to select AVX2/SSE implementation - but it doubles number of jobs anyway. This is not quite sustainable. 

I mean, I'm not 100% against it - but I'd like gains to be comparable. For example, if we rewrite more operations to vectorised instructions, it could become worth the hassle.

ethercrow commented:
> two CI servers: one with AVX2 and another without

I imagine it could be arranged that an environment with AVX2 pretends to not have it.

> This way we need to test builds against an older gcc, a newer gcc, an older clang and a newer clang.

I did something of the sort at a previous workplace using nix, it was pretty straightforward.

Bodigrim commented:
Practically speaking, to move forwards I'd like to see 
* a workaround for clang, 
* `#if __GNUC__ > 5` guard,
* some way to switch between SSE and AVX2 implementations. 

How much performance is gained by AVX2 version is comparison to SSE?

cartazio commented:
Any performance delta will be entirely dependent on which cpu architecture the sse or avx instructions run on. Eg on some amd stuff 256 ops underperform relative to 128. The agner fog cpu latency guide is an excellent and complete resource. 

Likewise, code that runs on Eg m1 Apple computers via x86_64 only support / emulate sse / pre avx instructions.

0xd34df00d commented:
@Bodigrim,

> Could your patch be rewritten in a more platform-independent way, possibly sacrificing some performance gains?

The only thing I can think of (that would still be worth the effort) is about using inline assembly for `cpuid` things as opposed to the intrinsic. Although it might turn out to not be that much more portable — it's been a while since I used inline assembly at all, so I might miss some differences between compilers wrt to that.

#310 is sort of lucky to be implementable with a reasonable speedup purely in terms of sse2, which is mandated by x86_64, so no extra checks needed.

> could you please rebase atop the latest master?

Will hopefully do that in a couple of days.

> For example, if we rewrite more operations to vectorised instructions, it could become worth the hassle.

That's a good point. I'm starting with `count` because that's the thing that I needed for my specific project. Other functions might also be rewritten once the need arises.

> How much performance is gained by AVX2 version is comparison to SSE?

As Carter said, that largely depends. I made some benchmarks a while ago on the hardware I have, so please feel free to <a href="https://github.com/0xd34df00d/counting-chars">take a look</a>. tl;dr 2x performance gains on Intel, largely insignificant ones on (recent) AMD.

Bodigrim commented:
@0xd34df00d I do not want this to linger indefinitely, especially in an outdated state with merge conflicts and red CI. If you do not have time/interest at least to rebase, I'll close this is stalled (with an option to reopen later again, of course).

> The only thing I can think of (that would still be worth the effort) is about using inline assembly for `cpuid` things as opposed to the intrinsic. 

No inline assembly please :)

0xd34df00d commented:
@Bodigrim sorry about that, I saw rebase conflicts and postponed fixing that, and then it kinda dropped out from my attention span.

Anyway, I've managed to:

1. rebase on top of current `master`, 
2. add a guard re gcc version,
3. add a guard workaround for clang (I still don't understand the root cause for why clang would reject a statically known `const int` as an argument for that intrinsic — at least, in my C++ experience that should have been a completely equivalent thing). Haven't checked that on macos or freebsd as I don't readily have those but it works on my linux machine.

Pushed just now, let's see if CI barks at me. Although, to be frank, the more I look at this the more I doubt if this stuff truly belongs to this package in its current form. Yet again, ideally I'd like to see all these functions (count, minimum, maximum, etc) implemented via ghc primops, but that would take a while. So, that's ultimately a choice between an immediately (sort of) available but error-prone solution vs something safer, but in the long-term — but that's up to bytestring maintainers. I don't have a strong opinion and I don't know what aligns with your plans or philosophy better.

Bodigrim commented:
Sorry, Windows build is flaky lately, but after two reruns CI is finally green :)
@ethercrow could you possibly review?

ethercrow commented on cbits/fpstring.c:
```diff
@@ -113,3 +107,160 @@ int fps_compare(const void *a, const void *b) {
+        __m128i p5 = _mm_load_si128((const __m128i*)(str + i + 16 * 5));
+        __m128i p6 = _mm_load_si128((const __m128i*)(str + i + 16 * 6));
+        __m128i p7 = _mm_load_si128((const __m128i*)(str + i + 16 * 7));
+#define MODE _SIDD_SBYTE_OPS | _SIDD_CMP_EQUAL_EACH
```
Please add the reference documentation link https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=1885,6090,5596,5653,4115,4076,4115,4938,4956,4115,5608,5656,5608,835,2426,835&text=_mm_cmpestrm for the next person who doesn't know what `_mm_cmpestrm` is.

ethercrow commented on cbits/fpstring.c:
```diff
@@ -113,3 +107,160 @@ int fps_compare(const void *a, const void *b) {
+        res += _popcnt64(_mm256_extract_epi64(r3, 3));
+    }
+
+    res /= 8;
```
I needed a moment to figure why this division is needed. Maybe add a comment that r0, r1, r2, r3 have 0xff (so 8 bits) for every matching byte and 0x00 for all the rest.

ethercrow commented:
> @ethercrow could you possibly review?

The code looks fine to me, I just have a couple of suggestions to add some explanations.

Bodigrim commented:
What I really like here (and what allows me to tolerate extra complexity) is that this patch makes `count` faster both for sparse and for dense chars. Currently a naive loop in `count` is slow when a character has a low frequency, but replacing it with `memchr` could cause a performance regression for high-frequency chars. It was even suggested earlier to provide two separate functions, which would be extremely awkward.

sjakobi commented on cbits/fpstring.c:
```diff
@@ -113,3 +107,160 @@ int fps_compare(const void *a, const void *b) {
+#ifndef USE_SIMD_COUNT
+    return fps_count_naive(str, len, w);
+#else
+    if (len <= 1024) {
```
A comment that explains this magic 1024 would be nice to have. Did you compare different cutoff values?

0xd34df00d replied:
Oh, I was waiting for somebody to ask! Nope, that's a rough guesstimate of how big the string should be for the SIMD stuff to compensate for all the extra work outside the main SIMD-based loop.

The optimal value will very much depend on the specific hardware this is going to be run on, and, moreover, for short strings (shorter than this cutoff) the absolute difference wouldn't be that big, and I presume the user just won't care. So, given that, I wouldn't put too much thought into it.

I can put a comment like `// a rough guesstimate` if this line of thought seems reasonable to you.

sjakobi replied:
Yes, please add a comment. :+1:

sjakobi reviewed (approved):
My C skills unfortunately are way too basic that I could provide an in-depth review. I think this is a great first step towards using SIMD instructions more widely in this code base though. :+1:

sjakobi commented:
Since this is such a novel thing in this code base, I'd love to see a review from @hsyl20 too.

cartazio commented:
I can do a secondary review. But it’s honestly solid clean simd code. With
a portable fallback

On Sat, Feb 13, 2021 at 1:27 PM Simon Jakobi <notifications@github.com>
wrote:

> Since this is such a novel thing in this code base, I'd love to see a
> review from @hsyl20 <https://github.com/hsyl20> too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/bytestring/pull/202#issuecomment-778658458>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABBQS4HAK2J5K32AK5AWTS63AAHANCNFSM4KO4U6PQ>
> .
>

Bodigrim commented:
> I can do a secondary review. But it’s honestly solid clean simd code. 

Feel free to approve ;) you are already on the list of reviewers.

> For portable robustness: it’d be great if somone could
> 1) test it on a currently supported arm platform
> 2) test it as Intel code on Apple m1

I believe arm is securely guarded by `#if defined(__x86_64__)` check.

0xd34df00d commented:
Thanks for the reviews! I'll address the feedback, plus I figured I want to look into adding more tests.

Re the M1 stuff (and ARM in general), an acquaintance of mine has ported the implementation to Neon as well, and it also turns out to be faster than what the compiler optimizes, and [here it is](https://github.com/Viknet/counting-chars/blob/neon/impls.h#L16-L59). I don't have any developer-friendly ARM hardware, so I cannot test this, but maybe somebody would be interested in incorporating that as well.

angerman commented:
> Thanks for the reviews! I'll address the feedback, plus I figured I want to look into adding more tests.
> 
> Re the M1 stuff (and ARM in general), an acquaintance of mine has ported the implementation to Neon as well, and it also turns out to be faster than what the compiler optimizes, and [here it is](https://github.com/Viknet/counting-chars/blob/neon/impls.h#L16-L59). I don't have any developer-friendly ARM hardware, so I cannot test this, but maybe somebody would be interested in incorporating that as well.

If someone adds this, and gives me some instructions on how to run the perf tests, I can run this on an M1 for comparison.

Bodigrim commented:
Let's pursue M1 thing in a separate PR, this thread is already year old ;)

@0xd34df00d could you please submit your benchmarks as well?

0xd34df00d commented:
Alright, addressed the PR comments (in a separate commit to simplify reviews; feel free to squash) and rebased over the current `master`.

Regarding the tests — I noticed there's already a QC property for the `count` function. IIRC there was a way to ensure QuickCheck tries bigger argument sizes as well, but what's the right way to do that with QC + Tasty? It seems reasonable to ensure QuickCheck tries big enough strings.

@Bodigrim , my original benchmarks were over a big `tmpfs`-residing `mmap`ed file. I'll throw up something more reproducible and add that to `bench`.

Bodigrim commented:
> It seems reasonable to ensure QuickCheck tries big enough strings.

The simplest thing I can come up with is something like 

```haskell
prop :: Char -> ByteString -> Int -> Property
prop ch xs n = count ch xs * n === count ch (replicate n xs)
```

so that `count` in the right-hand side is operating on long-ish strings.

ethercrow commented:
@Bodigrim I'd expect this property to cause OOM.

Bodigrim commented:
Because of too large `n`? One can use ``n `mod` 100000`` or similar.

0xd34df00d commented:
@Bodigrim got my hands on actually doing the benchmarks. Do you want to do just plain `count` one, or the one comparing naive native C implementation vs the SIMD-enabled one? I'm not sure the former is too helpful, and the latter would require some further work (namely, exposing the C implementation of the `fps_count_naive`, say, in an internal module, and benchmarking that function as well as the usual `count`).

Bodigrim commented:
> Do you want to do just plain `count` one, or the one comparing naive native C implementation vs the SIMD-enabled one?

A plain one, so that we can compare its performance before and after your branch.

Bodigrim commented on tests/Properties/ByteString.hs:
```diff
@@ -284,6 +284,11 @@ tests =
+  -- to ensure non-trivial prefix or suffix of the string is handled outside any possible SIMD-based loop,
+  -- which typically handles chunks of 16 or 32 or 64 etc bytes.
+  , testProperty "count (long strings)" $ mapSize (* 1025) $
+    \(toElem -> c) x -> B.length x >= 1024 ==> B.count c x === fromIntegral (length (elemIndices c (B.unpack x)))
```
I'm afraid this way a list in the right-hand size can quickly get out of hand. Imagine running tests with `--quickcheck-tests 100000`, and then these 100000 are multiplied by 1025. That's why I [was suggesting](https://github.com/haskell/bytestring/pull/202#issuecomment-786851139) a property with `replicate`.

cartazio commented:
👍

vdukhovni commented on cbits/fpstring.c:
```diff
@@ -90,19 +93,186 @@ unsigned char fps_minimum(unsigned char *p, size_t len) {
+        return fps_count_naive(str, len, w);
+    }
+
+    static _Thread_local fps_impl_t fps_simd_impl = NULL;
```
I've just spent a bunch of cycles dealing fallout related to _Thread_local use in <ctype.h> on FreeBSD.  I'd like to urge caution with their use in FFI.

The reason is that modules that use thread locals end up with relocations (TLSGD) that the RTS linker did not support at all (and with my most recent MR 5561, supports only if the thread-specific is an extern into an already loaded module).

What this means for bytestring is that it will no longer be possible to load bytestring into HLS or GHCI from a prebuilt .a archive.  There are no issues in fully compiled executables, but loading into a REPL or HLS is like to fail.

vdukhovni replied:
Have you looked at what sort of relocation (objdump -r on the ".a" archive) is produced when this code is compiled into a static package by GHC?  The static archives are JIT linked by the RTS linker in GHCI and HLS, and it is rather difficult to deal with TLS (thread-local-storage) linkage without help from the RTLD, but GHCI is loading ".o" files, not shared object modules.

cartazio commented:
If I’m understanding this correctly, the way to at least triage this issue is to have cabal default to building both static and dylib flavors and thus ghci can use Dylibs  and then the the issues are punted to folks who only build static lib flavors and use ghci?

vdukhovni commented:
> If I’m understanding this correctly, the way to at least triage this issue is to have cabal default to building both static and dylib flavors and thus ghci can use Dylibs and then the the issues are punted to folks who only build static lib flavors and use ghci?

In principle, yes, in practice, I haven't found this to be usable, because Cabal spats the dependencies across a large collection of subdirectories of the package-db, and there's no reasonable LD_LIBRARY_PATH one can use to make this go.

cartazio commented:
So to be clear: I think viktor is correct that certain config setups may break, but I think either those configs are wrong and tools that only work under those configurations should be fixed, or they don’t matter :). Which largely seem to relate to hls and ghci using its static linker magic 

I think the tls approach to writing this is flat out a pattern that is the only sane way to write this sort of portable but optimized ffi codes. And even if certain work flows get broken, the issue is they were already broken. And they won’t get fixed unless there’s a real reason to fix it. Let make sure things get better.  Let broken things learn and get fixed.

cartazio commented:
There’s totally a flat spot for shared Libs in the store folder ...   you might not see it if your .cabal/config file doesn’t enable dynamic flavor Libs

vdukhovni commented:
> I think the tls approach to writing this is flat out a pattern that is the only sane way to write this sort of portable but optimized ffi codes. And even if certain work flows get broken, the issue is they were already broken. And they won’t get fixed unless there’s a real reason to fix it. Let make sure things get better. Let broken things learn and get fixed.

I agree the tooling needs to get better, but breaking things first, with no fix in sight, and via a foundational package like bytestring is a bit too daring for me.  We should I think discuss this with @bgamari, @angerman, ... before we get too far ahead of the tooling...

cartazio commented:
See https://github.com/haskell/cabal/pull/4426

cartazio commented:
https://github.com/haskell/cabal/pull/4656

0xd34df00d commented:
I could take a look at storing the pointer in a single global variable (protected by a mutex or an rw lock) instead of a per-thread TLS. It shouldn't be too much of a performance overhead for any real-life scenario (especially given that the contention would be low). Let's see if I could make this not (further) break existing scenarios, even if they are already broken.

cartazio commented:
I’d personally not add thread safety nuance to this patch. Unless it’s super simple and has decent worst case overhead. I’d rather force stuff to get fixed

cartazio commented:
Or alternatively, if it’s simple enough, include both code paths and have tls be selected when we know it’ll work ? Like compiled code in an executable ?  
(i really want stuff to just work rather than accumulate work arounds for an issue we should address)

Bodigrim commented:
Sorry, I'm out of my depth here. What's the issue exactly? Is it reproducible on non-FreeBSD platforms?

vdukhovni commented:
> Sorry, I'm out of my depth here. What's the issue exactly? Is it reproducible on non-FreeBSD platforms?

Yes, it should be reproducible on most modern Elf systems.  The issue is that _Thread_local variables in "-fpic" ".o" files cannot be relocated by the GHC RTS linker. This means that "ghci" and "haskell-language-server", ... cannot runtime load static packages that use "_Thread_local" variables, unless (so far just on FreeBSD and only after 5561 is merged) the variable is "extern", e.g. a reference to a thread-local in libc.

vdukhovni commented:
> Yes, it should be reproducible on most modern Elf systems.

Provided that the toolchain tries to load the package from a static ".a" archive. rather than a ".so" file.  When both are available, quick tests show that "ghci" prefers the ".so", and "haskell-language-server" prefers the ".a".  So perhaps the impact will be limited to "hls" users.  I don't know why "hls" goes for the ".a" files, perhaps that's something to ask @mpickering.

cartazio commented:
If the only impacted users will be older hls builds because they bias to the static lib rather than dylib for dynamic code loading, it sounds like using thread local variables is fine and we should do it!

vdukhovni commented:
> If the only impacted users will be older hls builds because they bias to the static lib rather than dylib for dynamic code loading, it sounds like using thread local variables is fine and we should do it!

I just had a long chat with @bgamari and he explained that presently code that needs to be able repeated unload modules presently has no choice but to use the static linkage model, because dependency tracking in the RTS to support unloading only works in that case.  My impression is therefor that he'd not be keen to see TLS variables creeping into bytestring at this time, but I expect he'll also chime in for himself...

bgamari commented:
This all looks great. However, it does seem like this use of TLS will cause troubles, as noted by @vdukhovni. Specifically, it seems to be extremely difficult to handle TLS when loading static objects (as done by GHCi). The problems here are 

It's also not clear to me why TLS is needed here. Surely the desired `count` implementation isn't going to vary on a thread-by-thread basis. As such, we should only need to do this check once. Unfortunately, this is something that is [trivial](https://stackoverflow.com/questions/26955001/thread-safe-singleton-in-c11) to do in C++ but a bit trickier in C11 (requiring explicit atomics).

I believe the following would be sufficient:
```c
#include <stdatomic.h>
#include <stdbool.h>

bool check_sse_supported();

static bool sse_supported() {
  static atomic_int cache = -1;
  int supported = atomic_load_explicit(&cache, memory_order_relaxed);
  if (supported == -1) {
    supported = check_sse_supported();
    atomic_store_explicit(&cache, supported, memory_order_relaxed);
  }
  return cache == 1;
}
```

0xd34df00d commented:
Ok, implemented this with C11's [`call_once`](https://en.cppreference.com/w/c/thread/call_once). Although a bad libc implementation might have this function fairly unoptimized, it's only used for big enough strings, where the actual counting time much exceeds `call_once`. Although atomics almost certainly wouldt be faster, I'd much rather use a library function designed specifically for this task to avoid placing the burden of reasoning about atomics and the right memory orderings on myself and the next reader of the code.

vdukhovni commented:
Well, C11 `call_once` may not be terribly portable.  For example, on FreeBSD 12.2 it requires linking with `-lstdthreads`.  On Fedora 31 it requires linking with `-lpthread`.  Either way, it is not found in just `libc`.

On MacOS "Big Sur", there is no `<threads.h>` header file.

0xd34df00d commented:
I guess this means POSIX's `pthread_once` is also not an option. Welp, atomics it is then!

vdukhovni commented:
Well, pthread_once is an option on Unix-like systems, assuming explicit linking with `-lpthread` is not an obstacle. It is at least more broadly available, but then perhaps you run into an issue with Windows...

My opinion remains that these sort of primitives belong in GHC itself as new priomops, and not so much in `bytestring`, but it seems I'm a minority view on this...

Bodigrim commented:
Well, I've heard talks about vectorized operations in GHC for years without much fruition. Even if suddenly there emerges a champion to push this stuff forward, it would still take a long time before `bytestring` could use it. If we can speed up things without a (potentially infinite) delay, it's a right to do so. 

My personal opinion is that vectorized operations would not find way into GHC primitives until there accumulates a critical mass of libraries employing them. Then it would be easy to make a case in favor of.

0xd34df00d commented:
@vdukhovni I skimmed superficially through what GHC has on SIMD primops some time ago, and, while I more than agree with your sentiment for multiple reasons, I see several obstacles (in fact, I tried writing the above with primops, and stumbled upon a few missing ones).

Firstly, a more technical one: if I understand correctly, currently GHC only allows SIMD primops from `GHC.Prim` when building with the llvm codegen, which, in this case, means that `-fllvm` needs to be turned on when building `bytestring`. Of course, I'm sure that NCG can also be tought about the primops, so that's more of a technicality (that still needs to be accounted for, though).

Secondly, a more conceptual one: I'm not sure how to expose all the interesting instructions in an architecture-agnostic way. The usual things like vector addition are easy — they're present in pretty much any SIMD extension set with the same semantics. But if we are starting to consider things like [`cmpestrm`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestrm&expand=835), that becomes entirely non-obvious — I bet ARM doesn't have a direct substitute for this instruction.

One way to solve the latter is by allowing the user to define their own primops, that might look like
```haskell
cmpestrm# :: Word8x16# -> Word8x16# -> Imm8# -> Word8x16#
cmpestrm# s1 s2 imm8 = [ghcMagic| pcmpestrm s1 s2 imm8 |]
```
or something along those lines (where `Imm8#` stands for some immediate value known at compile-time).

Although I'm not sure the depth of the can of worms it opens: in fact, the above definition should also accept lengths that shall be stored in `eax`/`rax` and `edx`/`rdx` (that's the contract of `pcmpestrm`), so it's really a compound instruction, and the codegen needs to know those registers get clobbered.

Bodigrim commented:
Benchmarks look impressive, 70-85% faster for long strings! 
If there are no further suggestions (ping @hsyl20), I'll merge it in a couple of days.

vdukhovni commented:
> Secondly, a more conceptual one: I'm not sure how to expose all the interesting instructions in an architecture-agnostic way. The usual things like vector addition are easy — they're present in pretty much any SIMD extension set with the same semantics. But if we are starting to consider things like [`cmpestrm`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestrm&expand=835), that becomes entirely non-obvious — I bet ARM doesn't have a direct substitute for this instruction.

What I had in mind is not so much primops for non-portable low-level instructions, but in fact direct support for higher-level primitives (like memchr(), ...) in GHC via appropriate low-level primitives when available, or naive loops otherwise.  These should work not only for `ByteString` but also for `ByteArray`...

cartazio commented:
I agree with deadbeef, good simd api design is out of reach for ghc any
time soon.

On Sat, May 8, 2021 at 2:10 PM Viktor Dukhovni ***@***.***>
wrote:

> Secondly, a more conceptual one: I'm not sure how to expose all the
> interesting instructions in an architecture-agnostic way. The usual things
> like vector addition are easy — they're present in pretty much any SIMD
> extension set with the same semantics. But if we are starting to consider
> things like cmpestrm
> <https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestrm&expand=835>,
> that becomes entirely non-obvious — I bet ARM doesn't have a direct
> substitute for this instruction.
>
> What I had in mind is not so much primops for non-portable low-level
> instructions, but in fact direct support for higher-level primitives (like
> memchr(), ...) in GHC via appropriate low-level primitives when available,
> or naive loops otherwise. These should work not only for ByteString but
> also for ByteArray...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/bytestring/pull/202#issuecomment-835454317>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABBQVZSASYBQUOIWCOUPLTMV5ANANCNFSM4KO4U6PQ>
> .
>

Bodigrim commented:
Thanks, @0xd34df00d!

chrisdone commented:
I wonder about the performance of `count` combined with `par`, with the 1GB file split in half and processed in parallel. 350ms is a long time for my other cores to be doing nothing. 🤔

cartazio commented:
So if bytestring length is greater than say 100000 or 10_000 or something
 we should call a safe ffi version?

On Sun, May 16, 2021 at 4:42 PM Chris Done ***@***.***> wrote:

> I wonder about the performance of count combined with par, with the 1GB
> file split in half and processed in parallel. 350ms is a long time for my
> other cores to be doing nothing. 🤔
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/bytestring/pull/202#issuecomment-841874714>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABBQW2WKI6KEPQRP7PEO3TOAU2PANCNFSM4KO4U6PQ>
> .
>

vdukhovni commented:
> So if bytestring length is greater than say 100000 or 10_000 or something we should call a safe ffi version?
> […](#)
> On Sun, May 16, 2021 at 4:42 PM Chris Done ***@***.***> wrote: I wonder about the performance of count combined with par, with the 1GB file split in half and processed in parallel. 350ms is a long time for my other cores to be doing nothing. 🤔 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#202 (comment)](https://github.com/haskell/bytestring/pull/202#issuecomment-841874714)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAABBQW2WKI6KEPQRP7PEO3TOAU2PANCNFSM4KO4U6PQ> .

Wouldn't that conflict with the push the to us unpinned memory?

cartazio commented:
Nope.  It’s safe precisely because it’s pinned !

Alternatively, we could split it into a sequence of slices that yield back
to hs to call per slice.

My personal rule of thumb is to make sure ffi  things that take more than
10 microseconds get split up to play nicer with the scheduler

On Sun, May 16, 2021 at 9:39 PM Viktor Dukhovni ***@***.***>
wrote:

> So if bytestring length is greater than say 100000 or 10_000 or something
> we should call a safe ffi version?
> … <#m_3150912763288880980_>
> On Sun, May 16, 2021 at 4:42 PM Chris Done *@*.***> wrote: I wonder about
> the performance of count combined with par, with the 1GB file split in half
> and processed in parallel. 350ms is a long time for my other cores to be
> doing nothing. 🤔 — You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <#202 (comment)
> <https://github.com/haskell/bytestring/pull/202#issuecomment-841874714>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAABBQW2WKI6KEPQRP7PEO3TOAU2PANCNFSM4KO4U6PQ
> .
>
> Wouldn't that conflict with the push the to us unpinned memory?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/bytestring/pull/202#issuecomment-841920438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABBQSZBUE7D7PQ6KURZLLTOBXTJANCNFSM4KO4U6PQ>
> .
>

vdukhovni commented:
My impression was that "unsafe" calls are not preempted by the GC, and don't necessarily need pinned memory.  While "safe" calls do.  Since there are proposals in flight to move ByteString to unpinned ByteArray, I'd expect those to be less amenable for use with "safe" calls.

hsyl20 commented:
> If there are no further suggestions (ping @hsyl20), I'll merge it in a couple of days.

Sorry for not being very active these days (first baby at home is taking quite some time). I don't have much to add anyway. Good work! I also wish we had a better story for simd primops in GHC but we don't even have a good story for sub-word (Word8#, Word16€, Word32#...) yet so it will take some time.

cartazio commented:
I think bytestring won’t switch to unconditionally unpinned memory , but rather have pinned and unpinned flavors.  (It’s a deeply crash heavy breaking change to silently make them unpinned. May as well be new modules or a new package at that point :) )
